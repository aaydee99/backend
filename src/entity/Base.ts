import { PrimaryColumn, BaseEntity } from 'typeorm';
import { v4 as uuidv4 } from 'uuid';

export abstract class ExtendedBaseEntity extends BaseEntity {
	@PrimaryColumn('uuid')
	id = uuidv4();
}
