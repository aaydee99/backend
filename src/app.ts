import express from 'express';

import { useMiddlewares } from './middlewares';
import { useRoutes } from './routes';

const expressApp = express();
useMiddlewares(expressApp);
useRoutes(expressApp);

export const app = expressApp;
