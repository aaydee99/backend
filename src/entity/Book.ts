import { Entity, Column, BeforeInsert } from 'typeorm';

import { ExtendedBaseEntity } from './Base';

@Entity()
export class Book extends ExtendedBaseEntity {
	@Column('text', { unique: true })
	ISBN: string;

	@Column('text')
	title: string;

	@Column('text')
	subject: string;

	@Column('text')
	publisher: string;

	@Column()
	numberOfPages: number;

	@Column('text')
	author: string;

	@Column()
	numberOfCopies: number;

	@Column()
	numberOfAvailable: number;

	@Column('numeric', { precision: 2 })
	price: number;

	@BeforeInsert()
	setNumberOfAvailable() {
		this.numberOfAvailable = this.numberOfCopies;
	}
}
