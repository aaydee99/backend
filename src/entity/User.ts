import { hash } from 'bcryptjs';
import { Entity, Column, BeforeInsert, CreateDateColumn } from 'typeorm';

import { ExtendedBaseEntity } from './Base';

export enum UserStatus {
	ACTIVE = 'ACTIVE',
	CLOSED = 'CLOSED',
	CANCELED = 'CANCELED',
	BLACKLISTED = 'BLACKLISTED',
	NONE = 'NONE',
}

export enum UserRole {
	LIBRARIAN = 'LIBRARIAN',
	MEMBER = 'MEMBER',
}

@Entity('users')
export class User extends ExtendedBaseEntity {
	@Column('text')
	name: string;

	@Column('text', { unique: true })
	email: string;

	@Column('text')
	password: string;

	@Column('text')
	status: UserStatus = UserStatus.NONE;

	@Column('text')
	phone: string;

	@Column('text')
	role: UserRole = UserRole.MEMBER;

	@CreateDateColumn()
	dateOfMembership: Date;

	@Column()
	totalBooksCheckedout: number = 0;

	@Column()
	tokenVersion: number = 0;

	token: string;

	@BeforeInsert()
	async hashPass() {
		this.password = await hash(this.password, 10);
	}
}
