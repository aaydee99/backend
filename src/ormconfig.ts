import { ConnectionOptions } from 'typeorm';

import { DATABASE_URL, NODE_ENV } from './constants';

const ormConfig: ConnectionOptions[] = [
	{
		name: 'development',
		type: 'better-sqlite3',
		database: 'database.sqlite',
		synchronize: true,
		logging: true,
		// dropSchema: true,
		entities: ['src/entity/**/*.ts'],
		migrations: ['src/migration/**/*.ts'],
		subscribers: ['src/subscriber/**/*.ts'],
		cli: {
			entitiesDir: 'src/entity',
			migrationsDir: 'src/migration',
			subscribersDir: 'src/subscriber',
		},
	},
	{
		name: 'test',
		type: 'better-sqlite3',
		database: 'database.test.sqlite',
		synchronize: true,
		logging: false,
		dropSchema: true,
		entities: ['src/entity/**/*.ts'],
		migrations: ['src/migration/**/*.ts'],
		subscribers: ['src/subscriber/**/*.ts'],
		cli: {
			entitiesDir: 'src/entity',
			migrationsDir: 'src/migration',
			subscribersDir: 'src/subscriber',
		},
	},
	{
		name: 'production',
		type: 'postgres',
		url: DATABASE_URL,
		synchronize: true, // switch this to false after the initial tables created. use migrations instead afterwards
		logging: false,
		entities: ['dist/entity/**/*.js'],
		migrations: ['dist/migration/**/*.js'],
		subscribers: ['dist/subscriber/**/*.js'],
		cli: {
			entitiesDir: 'dist/entity',
			migrationsDir: 'dist/migration',
			subscribersDir: 'dist/subscriber',
		},
		ssl: true,
		extra: {
			ssl: {
				rejectUnauthorized: false,
			},
		},
	},
];

export const getConnectionConfig = (connName?: string) => {
	const config = ormConfig.find(
		(config) => (connName || NODE_ENV) === config.name
	);

	return config;
};
