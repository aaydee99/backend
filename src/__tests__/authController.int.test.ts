import request from 'supertest';
import { createConnection, getConnection } from 'typeorm';

import { app } from '../app';
import { User } from '../entity/User';
import { getConnectionConfig } from '../ormconfig';

interface ObjectWithUser {
	user: User;
}

const fetcher = request(app);
const dummyInput = {
	name: 'Coco',
	email: 'coco@gmail.com',
	password: 'Coco@2',
	phone: '09223456789',
};

let token = '';
const getAuthReqHeaders = () => {
	return ['Authorization', `Bearer ${token}`] as const;
};

beforeAll(async () => {
	const connectionConfig = getConnectionConfig();
	await createConnection({ ...connectionConfig!, name: 'default' });
});

afterAll(async () => {
	const connection = getConnection();
	await connection.close();
});

describe('POST_register controller tests', () => {
	it('validation works', async () => {
		const { text } = await fetcher
			.post('/auth/register')
			.set('Content-Type', 'application/json')
			.send({
				name: ' ',
				email: 'cocogmail.com',
				password: 'just5',
				phone: ' ',
			});

		expect(JSON.parse(text)).toMatchSnapshot();
	});

	it('can register not taken email', async () => {
		const { text } = await fetcher
			.post('/auth/register')
			.set('Content-Type', 'application/json')
			.send(dummyInput);

		const { user: resUser } = JSON.parse(text) as ObjectWithUser;
		const { id, dateOfMembership, token: jwt, ...user } = resUser;

		expect(id).toBeDefined();
		expect(dateOfMembership).toBeDefined();
		expect(jwt).toBeDefined();
		expect(user).toMatchSnapshot();
	});

	it("CAN'T register already taken email", async () => {
		const { text } = await fetcher
			.post('/auth/register')
			.set('Content-Type', 'application/json')
			.send(dummyInput);

		expect(JSON.parse(text)).toMatchSnapshot();
	});
});

describe('POST_login controller tests', () => {
	it('validation works', async () => {
		const { text } = await fetcher
			.post('/auth/login')
			.set('Content-Type', 'application/json')
			.send({
				email: 'cocogmail.com',
				password: '',
			});

		expect(JSON.parse(text)).toMatchSnapshot();
	});

	it("CAN'T login incorrect credentials", async () => {
		const { text } = await fetcher
			.post('/auth/login')
			.set('Content-Type', 'application/json')
			.send({
				...dummyInput,
				email: 'coco2@gmail.com',
			});

		expect(JSON.parse(text)).toMatchSnapshot();

		const { text: text2 } = await fetcher
			.post('/auth/login')
			.set('Content-Type', 'application/json')
			.send({
				...dummyInput,
				password: 'Coco@3',
			});

		expect(JSON.parse(text2)).toMatchSnapshot();
	});

	it('can login correct credentials', async () => {
		const { text } = await fetcher
			.post('/auth/login')
			.set('Content-Type', 'application/json')
			.send(dummyInput);

		const { user: resUser } = JSON.parse(text) as ObjectWithUser;
		const { id, dateOfMembership, token: jwt, ...user } = resUser;
		token = jwt;

		expect(id).toBeDefined();
		expect(dateOfMembership).toBeDefined();
		expect(jwt).toBeDefined();
		expect(user).toMatchSnapshot();
	});
});

describe('GET_me controller tests', () => {
	it('CAN\'T get "me" if not logged in', async () => {
		const { text } = await fetcher.get('/auth/me');
		expect(JSON.parse(text)).toMatchSnapshot();
	});

	it('can get "me" if logged in', async () => {
		const { text } = await fetcher.get('/auth/me').set(...getAuthReqHeaders());
		const { user: resUser } = JSON.parse(text) as ObjectWithUser;
		const { id, dateOfMembership, token: jwt, ...user } = resUser;

		expect(id).toBeDefined();
		expect(dateOfMembership).toBeDefined();
		expect(jwt).toBeDefined();
		expect(user).toMatchSnapshot();
	});
});

describe('GET_logout controller tests', () => {
	it("CAN'T logout if not logged in", async () => {
		const { text } = await fetcher.get('/auth/logout');
		expect(JSON.parse(text)).toMatchSnapshot();
	});

	it('can logout if logged in', async () => {
		const { text } = await fetcher
			.get('/auth/logout')
			.set(...getAuthReqHeaders());

		expect(JSON.parse(text)).toMatchSnapshot();

		const { text: text2 } = await fetcher
			.get('/auth/me')
			.set(...getAuthReqHeaders());

		expect(JSON.parse(text2)).toMatchSnapshot();
	});
});
