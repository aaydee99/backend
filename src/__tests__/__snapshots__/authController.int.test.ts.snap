// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET_logout controller tests CAN'T logout if not logged in 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "user is unauthorized",
      "path": "user",
    },
  ],
}
`;

exports[`GET_logout controller tests can logout if logged in 1`] = `
Object {
  "success": true,
}
`;

exports[`GET_logout controller tests can logout if logged in 2`] = `
Object {
  "errors": Array [
    Object {
      "message": "user is unauthorized",
      "path": "user",
    },
  ],
}
`;

exports[`GET_me controller tests CAN'T get "me" if not logged in 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "user is unauthorized",
      "path": "user",
    },
  ],
}
`;

exports[`GET_me controller tests can get "me" if logged in 1`] = `
Object {
  "email": "coco@gmail.com",
  "name": "Coco",
  "phone": "09223456789",
  "role": "MEMBER",
  "status": "NONE",
  "totalBooksCheckedout": 0,
}
`;

exports[`POST_login controller tests CAN'T login incorrect credentials 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "email or password is incorrect",
      "path": "email",
    },
  ],
}
`;

exports[`POST_login controller tests CAN'T login incorrect credentials 2`] = `
Object {
  "errors": Array [
    Object {
      "message": "email or password is incorrect",
      "path": "email",
    },
  ],
}
`;

exports[`POST_login controller tests can login correct credentials 1`] = `
Object {
  "email": "coco@gmail.com",
  "name": "Coco",
  "phone": "09223456789",
  "role": "MEMBER",
  "status": "NONE",
  "totalBooksCheckedout": 0,
}
`;

exports[`POST_login controller tests validation works 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "email must be a valid email",
      "path": "email",
    },
    Object {
      "message": "password is a required field",
      "path": "password",
    },
  ],
}
`;

exports[`POST_register controller tests CAN'T register already taken email 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "email is already registered",
      "path": "email",
    },
  ],
}
`;

exports[`POST_register controller tests can register not taken email 1`] = `
Object {
  "email": "coco@gmail.com",
  "name": "Coco",
  "phone": "09223456789",
  "role": "MEMBER",
  "status": "NONE",
  "totalBooksCheckedout": 0,
}
`;

exports[`POST_register controller tests validation works 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "name is a required field",
      "path": "name",
    },
    Object {
      "message": "email must be a valid email",
      "path": "email",
    },
    Object {
      "message": "password must be at least 6 characters",
      "path": "password",
    },
    Object {
      "message": "phone is a required field",
      "path": "phone",
    },
  ],
}
`;
